src/libproc/proc_pid.rs

```diff
@@ -637,3 +637,452 @@ impl From<uint32_t> for ProcFDType {
     }
 }
 
+// This trait is needed for polymorphism on pidfdinfo types, also abstracting flavor in order to provide
+// type-guaranteed flavor correctness
+pub trait PIDFDInfo: Default {
+    fn flavor() -> PidFDInfoFlavor;
+}
+
+/// Returns the information about file descriptors of the process that match pid passed in.
+///
+/// # Examples
+///
+/// ```
+/// use std::io::Write;
+/// use std::net::TcpListener;
+/// use libproc::libproc::proc_pid::{listpidinfo, pidinfo, pidfdinfo, ListFDs, ListThreads, BSDInfo, ProcFDType, SocketFDInfo, SocketInfoKind};
+///
+/// fn pidfdinfo_test() {
+///     use std::process;
+///     let pid = process::id() as i32;
+///
+///     // Open TCP port:8000 to test.
+///     let _listener = TcpListener::bind("127.0.0.1:8000");
+///
+///     if let Ok(info) = pidinfo::<BSDInfo>(pid, 0) {
+///         if let Ok(fds) = listpidinfo::<ListFDs>(pid, info.pbi_nfiles as usize) {
+///             for fd in &fds {
+///                 match fd.proc_fdtype.into() {
+///                     ProcFDType::Socket => {
+///                         if let Ok(socket) = pidfdinfo::<SocketFDInfo>(pid, fd.proc_fd) {
+///                             match socket.psi.soi_kind.into() {
+///                                 SocketInfoKind::Tcp => {
+///                                     // access to the member of `soi_proto` is unsafe becasuse of union type.
+///                                     let info = unsafe { socket.psi.soi_proto.pri_tcp };
+///
+///                                     // change endian and cut off because insi_lport is network endian and 16bit witdh.
+///                                     let mut port = 0;
+///                                     port |= info.tcpsi_ini.insi_lport >> 8 & 0x00ff;
+///                                     port |= info.tcpsi_ini.insi_lport << 8 & 0xff00;
+///
+///                                     // access to the member of `insi_laddr` is unsafe becasuse of union type.
+///                                     let s_addr = unsafe { info.tcpsi_ini.insi_laddr.ina_46.i46a_addr4.s_addr };
+///
+///                                     // change endian because insi_laddr is network endian.
+///                                     let mut addr = 0;
+///                                     addr |= s_addr >> 24 & 0x000000ff;
```

Comment:
I've never used it, but maybe rust has something to offer you to make endianness related stuff more idiomatic?
https://docs.rs/byteorder/1.3.1/byteorder/ maybe?